ext/intl/grapheme/grapheme_string.c

```diff
@@ -918,4 +919,216 @@ PHP_FUNCTION(grapheme_str_split)
+		Z_PARAM_STR(string2)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG(cost_ins)
+		Z_PARAM_LONG(cost_rep)
+		Z_PARAM_LONG(cost_del)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (cost_ins <= 0 || cost_ins > UINT_MAX / 4) {
+		zend_argument_value_error(3, "must be greater than 0 and less than or equal to %d", UINT_MAX / 4);
+		RETURN_THROWS();
+	}
+
+	if (cost_rep <= 0 || cost_rep > UINT_MAX / 4) {
+		zend_argument_value_error(4, "must be greater than 0 and less than or equal to %d", UINT_MAX / 4);
+		RETURN_THROWS();
+	}
+
+	if (cost_del <= 0 || cost_del > UINT_MAX / 4) {
+		zend_argument_value_error(5, "must be greater than 0 and less than or equal to %d", UINT_MAX / 4);
+		RETURN_THROWS();
+	}
+
+	zend_long *p1, *p2, *tmp;
+	zend_long c0, c1, c2;
+	zend_long retval;
+	size_t i2;
+	char *pstr1, *pstr2;
+
+	UChar *ustring1 = NULL;
+	UChar *ustring2 = NULL;
+
+	int32_t ustring1_len = 0;
+	int32_t ustring2_len = 0;
+
+	UErrorCode ustatus1 = U_ZERO_ERROR;
+	UErrorCode ustatus2 = U_ZERO_ERROR;
+
+	/* When all costs are equal, levenshtein fulfills the requirements of a metric, which means
+	 * that the distance is symmetric. If string1 is shorter than string2 we can save memory (and CPU time)
+	 * by having shorter rows (p1 & p2). */
+	if (ZSTR_LEN(string1) < ZSTR_LEN(string2) && cost_ins == cost_rep && cost_rep == cost_del) {
+		zend_string *tmp = string1;
+		string1 = string2;
+		string2 = tmp;
+	}
+
+	pstr1 = ZSTR_VAL(string1);
+	pstr2 = ZSTR_VAL(string2);
+
+	intl_convert_utf8_to_utf16(&ustring1, &ustring1_len, pstr1, ZSTR_LEN(string1), &ustatus1);
+
+	if (U_FAILURE(ustatus1)) {
+		intl_error_set_code(NULL, ustatus1);
+
+		intl_error_set_custom_msg(NULL, "Error converting input string to UTF-16", 0);
+		efree(ustring1);
+		RETURN_FALSE;
+	}
+
+	intl_convert_utf8_to_utf16(&ustring2, &ustring2_len, pstr2, ZSTR_LEN(string2), &ustatus2);
+
+	if (U_FAILURE(ustatus2)) {
+		intl_error_set_code(NULL, ustatus2);
+
+		intl_error_set_custom_msg(NULL, "Error converting input string to UTF-16", 0);
+		efree(ustring2);
+		efree(ustring1);
+		RETURN_FALSE;
+	}
+
+	UBreakIterator *bi1, *bi2;
+
+	int32_t strlen_1, strlen_2;
+	strlen_1 = grapheme_split_string(ustring1, ustring1_len, NULL, 0);
+	strlen_2 = grapheme_split_string(ustring2, ustring2_len, NULL, 0);
+
+	if (strlen_1 == 0) {
+		efree(ustring1);
+		efree(ustring2);
+		RETURN_LONG(strlen_2 * cost_ins);
+	}
+	if (strlen_2 == 0) {
+		efree(ustring1);
+		efree(ustring2);
+		RETURN_LONG(strlen_1 * cost_del);
+	}
+
+	unsigned char u_break_iterator_buffer1[U_BRK_SAFECLONE_BUFFERSIZE];
+	unsigned char u_break_iterator_buffer2[U_BRK_SAFECLONE_BUFFERSIZE];
+	bi1 = grapheme_get_break_iterator((void*)u_break_iterator_buffer1, &ustatus1);
+	bi2 = grapheme_get_break_iterator((void*)u_break_iterator_buffer2, &ustatus2);
+
+	ubrk_setText(bi1, ustring1, ustring1_len, &ustatus1);
+
+	if (U_FAILURE(ustatus1)) {
+		intl_error_set_code(NULL, ustatus1);
+
+		intl_error_set_custom_msg(NULL, "Error on ubrk_setText on ustring1", 0);
```

Comment:
The error message could probably be improved (same below)
e.g. `"Error on ubrk_setText for argument #1 ($string1)"`